awx/api/serializers.py

```diff
@@ -3375,22 +3422,57 @@ def validate(self, attrs):
         JT_job_tags = obj.job_tags
         JT_skip_tags = obj.skip_tags
         JT_inventory = obj.inventory
-        JT_credential = obj.credential
         JT_verbosity = obj.verbosity
-        extra_credentials = attrs.pop('extra_credentials', None)
+        credentials = attrs.pop('credentials', None)
         attrs = super(JobLaunchSerializer, self).validate(attrs)
         obj.extra_vars = JT_extra_vars
         obj.limit = JT_limit
         obj.job_type = JT_job_type
         obj.skip_tags = JT_skip_tags
         obj.job_tags = JT_job_tags
         obj.inventory = JT_inventory
-        obj.credential = JT_credential
         obj.verbosity = JT_verbosity
-        if extra_credentials is not None:
-            attrs['extra_credentials'] = extra_credentials
+        if credentials is not None:
+            attrs['credentials'] = credentials
+
+        # fill passwords dict with request data passwords
+        for cred in (attrs['credentials'] or obj.credentials.all()):
+            if cred.passwords_needed:
+                passwords = self.context.get('passwords')
+                try:
+                    for p in cred.passwords_needed:
+                        passwords[p] = data[p]
+                except KeyError:
+                    errors.setdefault('passwords_needed_to_start', []).extend(cred.passwords_needed)
```

Comment:
If you add to `errors` here, it won't raise the validation error since it's after

```python
         if errors:
             raise serializers.ValidationError(errors)
```

Reply:
You're right.  And the test I wrote to test this is wrong and misses it.  Good catch :)

Reply:
Hrm, you're right.  This validation *is* being properly handled, though.  It must be coming from elsewhere...